### PR TITLE
Prepare the rewriting step for using conditions.

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
The tests for Step/Function/Evaluator.hs, Step/Function/UserDefined.hs, Step/Substitution.hs, Unification.UnifierImpl.hs are still incomplete, but I'll probably finish them tomorrow and this PR is otherwise ready for review.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance

---

### CODE REVIEW

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+{--| 'PredicateProof' is a placeholder for a proof showing that a Predicate
+evaluation was correct.
+--}
+data PredicateProof level = PredicateProof
+    deriving (Show, Eq)
+
+{--| 'GenericPredicate' is a wrapper for predicates used for type safety.
+Should not be exported, and should be treated as an opaque entity which
+can be manipulateds only by functions in this module.
```
Comment (traiansf) [inline src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: manipulated*s*

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+{--| 'CommonPredicate' follows the generic convention of particularizing types
+to Variable.
+--}
+type CommonPredicate level = Predicate level Variable
+
+{- 'compactPredicatePredicate' removes one level of 'GenericPredicate' which
+sometimes occurs when, say, using Predicates as Traversable.
+-}
+compactPredicatePredicate
```
Comment (traiansf) [inline src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: This is an instance of join from Monad;  and the way you use it seems to be similar to how join is used...

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    then GenericPredicate patt
+    else error "Trying to wrap non-predicate as predicate!"
+
+{- 'unwrapPredicate' wraps a pattern in a GenericPredicate. This should be
+not be used outside of that.
+
+We should consider deleting this and implementing the functionality otherwise.
+-}
+unwrapPredicate :: Predicate level var -> PureMLPattern level var
+unwrapPredicate (GenericPredicate p) = p
```
Comment (traiansf) [inline src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: Should we have an `impl` module which contains all these relevant for testing and only export the smaller / relevant API from the non-impl module?

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+makeMultipleAndPredicate =
+    foldl'
+        (\(cond1, _) cond2 -> makeAndPredicate cond1 cond2)
+        (makeTruePredicate, PredicateProof)
+
+{--| 'makeAndPredicate' combines two Predicates with an 'and', doing some
+simplification.
+--}
+makeAndPredicate
+    -- TODO(virgil): Group these constraints in a class.
```
Comment (traiansf) [inline src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: Why not grouping these in a class, or even better a type (like `ShowMetaOrObject` in `MetaOrObject`), now?

File: src/main/haskell/language-kore/src/Data/Kore/Step/ExpandedPattern.hs [hunk 5]
```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE PatternSynonyms       #-}
+module Data.Kore.Step.ExpandedPattern
```
Comment (traiansf) [inline src/main/haskell/language-kore/src/Data/Kore/Step/ExpandedPattern.hs]: Module header

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs (hunk 1 above)
Comment (virgil-serbanuta) [reply src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: Fixed.

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs (hunk 2 above)
Comment (virgil-serbanuta) [reply src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: I wouldn't make this a Monad just to have a nice join.

File: src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs (hunk 3 above)
Comment (virgil-serbanuta) [reply src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs]: This is also used in the evaluation package.

File: src/main/haskell/language-kore/src/Data/Kore/Step/ExpandedPattern.hs (hunk 5 above)
Comment (virgil-serbanuta) [reply src/main/haskell/language-kore/src/Data/Kore/Step/ExpandedPattern.hs]: Done.

**Pull request merged**